src/game/client/shadow/w3dvolumetricshadow.cpp

```diff
@@ -0,0 +1,503 @@
+    geo_mesh->m_numPolygons = model->Get_Polygon_Count();
+    geo_mesh->m_polygons = model->Get_Polygon_Array();
+
+    if (geo_mesh->m_numVerts > 16384) {
+        return 0;
+    }
+
+    unsigned short verts[16384];
+    memset(verts, -1, sizeof(verts));
+
+    int num = geo_mesh->m_numVerts;
+
+    for (int i = 0; i < geo_mesh->m_numVerts; ++i) {
+        if (verts[i] == 0xFFFF) {
+
+            for (int j = i + 1; j < geo_mesh->m_numVerts; ++j) {
+
+                Vector3 d = geo_mesh->m_verts[i] - geo_mesh->m_verts[j];
+
+                if (d.Length2() == 0.0f) {
+                    verts[j] = i;
+                    --num;
+                }
+            }
+            verts[i] = i;
+        }
+    }
+
+    geo_mesh->m_parentVerts = new unsigned short[geo_mesh->m_numVerts];
+    memcpy(geo_mesh->m_parentVerts, verts, geo_mesh->m_numVerts * sizeof(unsigned short));
+
+    geo_mesh->m_numVerts = num;
+    geo_mesh->m_parentGeometry = this;
+
+    m_meshCount = 1;
+    m_numTotalsVerts = num;
+
+    return 1;
+}
+
+W3DShadowGeometryManager::W3DShadowGeometryManager()
+{
+    m_geomPtrTable = new HashTableClass(2048);
+    m_missingGeomTable = new HashTableClass(2048);
+}
+
+W3DShadowGeometryManager::~W3DShadowGeometryManager()
+{
+    Free_All_Geoms();
+}
+
+void W3DShadowGeometryManager::Free_All_Geoms()
+{
+    W3DShadowGeometryManagerIterator it(*this);
+
+    for (it.First(); !it.Is_Done(); it.Next()) {
+        W3DShadowGeometry *v1 = it.Get_Current_Geom();
+        v1->Release_Ref();
+    }
+
+    m_geomPtrTable->Reset();
+}
+
+W3DShadowGeometry *W3DShadowGeometryManager::Peek_Geom(const char *name)
+{
+    return (W3DShadowGeometry *)m_geomPtrTable->Find(name);
+}
+
+W3DShadowGeometry *W3DShadowGeometryManager::Get_Geom(const char *name)
+{
+    W3DShadowGeometry *geo = Peek_Geom(name);
+
+    if (geo != nullptr) {
+        geo->Add_Ref();
+    }
+
+    return geo;
+}
+
+bool W3DShadowGeometryManager::Add_Geom(W3DShadowGeometry *new_geom)
+{
+
+    captainslog_assert(new_geom != nullptr);
+
+    new_geom->Add_Ref();
+    m_geomPtrTable->Add(new_geom);
+
+    return true;
+}
+
+void W3DShadowGeometryManager::Register_Missing(const char *name)
+{
+    m_missingGeomTable->Add(new MissingGeomClass(name));
+}
+
+bool W3DShadowGeometryManager::Is_Missing(const char *name) const
+{
+    return m_missingGeomTable->Find(name) != nullptr;
+}
+
+int W3DShadowGeometryManager::Load_Geom(RenderObjClass *robj, const char *name)
+{
+    // TODO could be cleaner
+
+    int state = 0;
+
+    W3DShadowGeometry *geo = new W3DShadowGeometry;
+
+    if (geo != nullptr) {
+        geo->Set_Name(name);
+
+        int id = robj->Class_ID();
+
+        if (id == RenderObjClass::CLASSID_MESH) {
+            state = geo->Init_From_Mesh(robj) != 0;
+        } else if (id == RenderObjClass::CLASSID_HLOD) {
+            state = geo->Init_From_HLOD(robj) != 0;
+        }
+
+        if (state != 1) {
+            geo->Release_Ref();
```

Comment:
Something is not right here. First of all if code would reach here, application would crash a little further down below, because `geo` will be released (deleted) twice.

Second, `state` looks like it should be a `bool`?

Reply:
agreed release_ref is only called once in the assembly